### PR TITLE
PYTHON-2858 Use OP_MSG to authenticate if server supports OP_MSG

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
Is this ready for review?

File: pymongo/pool.py [hunk 1]
```diff
@@ -552,7 +552,8 @@ def unpin(self):
     def hello_cmd(self):
         # Handshake spec requires us to use OP_MSG+hello command for the
         # initial handshake in load balanced or versioned api mode.
-        if self.opts.server_api or self.hello_ok or self.opts.load_balanced:
+        if (self.opts.server_api or self.hello_ok or self.opts.load_balanced
+            or self.max_wire_version >= 6):
```
Comment (ShaneHarvey) [inline pymongo/pool.py:556]: This change isn't needed. We already set op_msg_enabled to true when `max_wire_version >= 6)` elsewhere.

File: test/mockupdb/test_handshake.py [hunk 2]
```diff
@@ -84,16 +85,13 @@ def test_client_handshake_data(self):
             self.addCleanup(server.stop)
 
         hosts = [server.address_string for server in (primary, secondary)]
-        primary_response = OpReply('ismaster', True,
-                                   setName='rs', hosts=hosts,
-                                   minWireVersion=2, maxWireVersion=6)
-        error_response = OpReply(
+        primary_response = {"setName": 'rs', "hosts": hosts,
+                            "minWireVersion": 2, "maxWireVersion": 13}
+        error_response = OpMsgReply(
```
Comment (ShaneHarvey) [inline test/mockupdb/test_handshake.py:90]: This test is already passing so do we need to change it at all?

File: test/mockupdb/test_handshake.py [hunk 3]
```diff
@@ -217,5 +220,42 @@ def test_handshake_not_either(self):
...
+                                         **{'payload':
+                                                b'r=wPleNM8S5p8gMaffMDF7Py4ru9bnmmoqb0'
+                                                b'1WNPsil6o=pAvr6B1garhlwc6MKNQ93ZfFky'
+                                                b'tXdF9r,'
+                                                b's=4dcxugMJq2P4hQaDbGXZR8uR3ei'
+                                                b'PHrSmh4uhkg==,i=15000',
+                                            "saslSupportedMechs": [
+                                                "SCRAM-SHA-1"]}))
+            else:
+                return request.reply(OpMsgReply(**primary_response))
```
Comment (ShaneHarvey) [inline test/mockupdb/test_handshake.py:243]: This looks like it responds to an OP_QUERY ismaster handshake with an OP_MSG response. Let's change it to use OP_REPLY. Eg:  `request.reply(**primary_response)`

File: test/mockupdb/test_handshake.py [hunk 4]
```diff
@@ -217,5 +220,42 @@ def test_handshake_not_either(self):
...
+
+        server.autoresponds(responder)
+        self.addCleanup(server.stop)
+        server.run()
+        client = MongoClient(server.uri,
+                             username='username',
+                             password='password',
+                             appname='my app',
```
Comment (ShaneHarvey) [inline test/mockupdb/test_handshake.py:251]: appname is not needed.

File: test/mockupdb/test_handshake.py [hunk 5]
```diff
@@ -217,5 +220,42 @@ def test_handshake_not_either(self):
...
+                             )
+        self.addCleanup(client.close)
+        self.assertRaises(OperationFailure, client.db.collection.find_one,
+                          {"a": 1})
+        assert self.found_auth_msg, """Could not find authentication command
```
Comment (ShaneHarvey) [inline test/mockupdb/test_handshake.py:256]: Use self.assertTrue with regular string (not triple quotes).

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/pool.py (hunk 1 above)
Comment (juliusgeo) [reply pymongo/pool.py]: Oops, yeah. You're correct. This was due to me confusing [DRIVERS-1916](https://jira.mongodb.org/browse/DRIVERS-1916) with saying that OP_MSG necessitates "hello" not the other way around. Reverted.

File: test/mockupdb/test_handshake.py (hunk 3 above)
Comment (juliusgeo) [reply test/mockupdb/test_handshake.py]: Done.

File: test/mockupdb/test_handshake.py (hunk 4 above)
Comment (juliusgeo) [reply test/mockupdb/test_handshake.py]: Done.

File: test/mockupdb/test_handshake.py (hunk 5 above)
Comment (juliusgeo) [reply test/mockupdb/test_handshake.py]: Done.

**Review by ShaneHarvey — COMMENTED** [review]
Can you schedule the mockupdb test suite?

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

**Review by ShaneHarvey — APPROVED** [review]
Still LGTM

**Pull request merged**